### DISCUSSION TITLE
relation between .chezmoiremove and .chezmoiignore

### DISCUSSION THREAD

**Discussion by kdomanski** [twpayne/chezmoi#1789] [Q&A] [answered]
Hi,
I noticed that a file marked for deletion in `.chezmoiremove` will not be removed if it is also a match in `.chezmoiignore`. Is that intentional? If it is, then how would I make such a scenario happen:
- `if .condition` create/update the file
- `if not .condition` do not create the file, and remove it if it already exists

**Comment by kdomanski** [comment, marked as answer]
Nevermind, I just found out in docs:

```
If, after executing the template, the file contents are empty, the target file will be removed.
```

**Comment by felipecrs** [comment]
I have the same problem. If the file is in `.chezmoiignore`, it's not being removed when in `.chezmoiremove`:

```console
❯ stat /etc/apt/sources.list.d/communitheme-ppa.list
  File: /etc/apt/sources.list.d/communitheme-ppa.list
  Size: 0               Blocks: 0          IO Block: 4096   regular empty file
Device: 820h/2080d      Inode: 188         Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2022-01-06 14:15:15.987294439 -0300
Modify: 2022-01-06 14:15:15.987294439 -0300
Change: 2022-01-06 14:15:15.987294439 -0300
 Birth: -
❯ cat /home/felipecrs/.dotfiles/root/.chezmoiremove | chezmoi execute-template
etc/apt/sources.list.d/git-core-ubuntu-ppa-*.list*
etc/apt/sources.list.d/communitheme-ubuntu-ppa-*.list*
etc/apt/sources.list.d/devel:kubic:libcontainers:stable.list*


etc/apt/sources.list.d/communitheme-ppa.list
usr/share/keyrings/communitheme-ppa-archive-keyring.gpg

❯ cat /home/felipecrs/.dotfiles/root/.chezmoiignore | chezmoi execute-template

etc/apt/sources.list.d/communitheme-ppa.list



❯ chezmoi status
 R .chezmoiscripts/20-install-prerequisites.sh
 R .chezmoiscripts/30-uninstall-apt-packages.sh
 R .chezmoiscripts/90-restore-permissions.sh
MM usr/local/share/gcm-core/git-credential-manager-core.exe
 R .chezmoiscripts/10-install-apt-packages.sh
 R .chezmoiscripts/20-configure-user.sh
 R .chezmoiscripts/30-fix-permissions.sh
 R .chezmoiscripts/90-fix-chezmoi-files-ownership.sh

```

I just wonder if this is a bug, or if it's per design.

**Reply by twpayne** [reply]
It's by design.

`.chezmoiignore` is for files that chezmoi should ignore completely, i.e. for files that chezmoi should *not* manage.

`.chezmoiremove` is for files that chezmoi should remove, i.e. for files that chezmoi should manage by removing them.

One of `ignore` or `remove` has to take priority. To me it seems more logical that `ignore` should take priority over `remove`. Both files are templates so you can tune them to match your personal configuration.

**Reply by felipecrs** [reply]
Thank you for the explanation. So, perhaps one way to handle it is:

```
$ cat .chezmoiignore
{{ if or (not .is_gnome) (stat "/etc/apt/sources.list.d/communitheme-ppa.list") }}
etc/apt/sources.list.d/communitheme-ppa.list
{{ end }}

$ cat .chezmoiremove
{{ if not .is_gnome }}
etc/apt/sources.list.d/communitheme-ppa.list
{{ end }}
```

This way, it will not only create the file when not running in a Gnome environment, but it will also delete in case it's there.

It's just a bit boring to do this when you have a big amount of files.

**Reply by felipecrs** [reply]
EDIT: that does not work. If I leave it in both `.chezmoiignore` and `.chezmoiremove`, it will trigger an inconsistent state error.

Is there a method for solving this situation without using scripts?

**Reply by felipecrs** [reply]
EDIT 2: actually, it does work. My bad. Sorry.

**Reply by felipecrs** [reply]
EDIT 3: it does not work indeed, as the file would have been declared both in my source state as a regular file and in `.chezmoiremove`, in case it's not in `.chezmoiignore`.

**Reply by felipecrs** [reply]
I implemented a workaround using scripts to handle this limitation of `.chezmoiremove` at https://github.com/felipecrs/dotfiles/commit/89666801f53cf1e6497665cc615e40878ead6b02.

I noticed that I was relying on this behavior since a long time for more places, and I was erroneously thinking that `chezmoi` was taking care of the removal.

If in the future we happen to have a chezmoi native solution for this situation, I would be happy to remove this workaround from my dotfiles.
